Add mnmg deployment details for AzureML

skirui-source opened the pull request:
Fixes https://github.com/rapidsai/deployment/issues/15 (refer to [this comment](https://github.com/rapidsai/deployment/issues/15#issuecomment-1461714779))

- [x] How to [launch a notebook](https://learn.microsoft.com/en-us/azure/machine-learning/quickstart-run-notebooks) with RAPIDS (SNSG/SNMG).
- [x] How to use the [native AzureML training jobs](https://learn.microsoft.com/en-us/azure/machine-learning/how-to-train-model?tabs=python) to launch a RAPIDS cluster (MNMG).
- [ ] Further reading pointing to [Azure VM](https://docs.rapids.ai/deployment/stable/cloud/azure/azure-vm-multi.html) and [AKS](https://docs.rapids.ai/deployment/stable/cloud/azure/aks.html) as alternatives to running an AzureML training job.

skirui-source commented:
@jacobtomlinson I think this is ready for initial review/comments, Is it detailed enough?
I was trying to avoid copying too much demo-code since I will be linking the notebook example anyway,

Lmk what you think!

jacobtomlinson commented on source/cloud/azure/azure-ml-multi-node.md:
````diff
@@ -0,0 +1,72 @@
+
+```console
+# Get a handle to the workspace
+ml_client = MLClient(
````
Where do I import `MLClient` from?

jacobtomlinson replied:
Using the `rapids` environment in the compute instance I had to run

```python
!pip install azure-ai-ml
from azure.ai.ml import MLClient
from azure.identity import DefaultAzureCredential
```

skirui-source replied:
apologies, i will include the imports in the snippet...

jacobtomlinson commented on source/cloud/azure/azure-ml-multi-node.md:
```diff
@@ -0,0 +1,72 @@
+
+### Instantiate workspace
+
+If using the Python SDK, connect to your workspace either by explicitly providing the workspace details or load from `config.json` file.
```
Ah this is the point where we use the `config.json` file that is mentioned on the single node page.

jacobtomlinson replied:
I don't actually see anywhere on the workspace page to download this file though (hope you don't mind me reusing your workspace). Am I missing it?

<img width="1180" alt="image" src="https://user-images.githubusercontent.com/1610850/230330673-5b6b16c4-34ea-4231-b0cb-3a7cef27e685.png">

skirui-source replied:
the `config.json` file is downloaded from the workspace [homepage](ttps://portal.azure.com/?feature.msaljs=true#@NVIDIA.onmicrosoft.com/resource/subscriptions/fc4f4a6b-4041-4b1c-8249-854d68edcf62/resourceGroups/rapidsai-deployment/providers/Microsoft.MachineLearningServices/workspaces/rapids-aml-cluster/overview) in Azure ML portal (not the Studio)

jacobtomlinson commented on source/cloud/azure/azure-ml-multi-node.md:
````diff
@@ -0,0 +1,72 @@
+You may choose to use low-priority VMs to run your workloads. These VMs don't have guaranteed availability but allow you to take advantage of Azure's unused capacity at a significant cost savings. The amount of available capacity can vary based on size, region, time of day, and more.
+
+```console
+gpu_compute = AmlCompute(
````
Where do I import `AMLCompute` from?

jacobtomlinson replied:
Found it `from azure.ai.ml.entities import AmlCompute`.

jacobtomlinson commented on source/cloud/azure/azure-ml-multi-node.md:
```diff
@@ -0,0 +1,72 @@
+
+`max_instances`: The max number of nodes to autoscale up to when you run a job
+
+NOTE:
+You may choose to use low-priority VMs to run your workloads. These VMs don't have guaranteed availability but allow you to take advantage of Azure's unused capacity at a significant cost savings. The amount of available capacity can vary based on size, region, time of day, and more.
```
This should be a note block
````suggestion
```{note}
You may choose to use low-priority VMs to run your workloads. These VMs don't have guaranteed availability but allow you to take advantage of Azure's unused capacity at a significant cost savings. The amount of available capacity can vary based on size, region, time of day, and more.
```
````

jacobtomlinson commented on source/cloud/azure/azure-ml-multi-node.md:
````diff
@@ -0,0 +1,72 @@
+ml_client.begin_create_or_update(gpu_compute).result()
+```
+
+### Upload data to Datastore
````
This section appears incomplete.

jacobtomlinson commented on source/cloud/azure/azure-ml-multi-node.md:
```diff
@@ -0,0 +1,72 @@
+    idle_time_before_scale_down=300,
+    tier="low_priority", # optional
+)
+ml_client.begin_create_or_update(gpu_compute).result()
```
Once I run this it's unclear what I need to do. How do I use the cluster? How do I delete it again?

I see the following output in the notebook.

```
AmlCompute({'type': 'amlcompute', 'created_on': None, 'provisioning_state': 'Succeeded', 'provisioning_errors': None, 'name': 'jtomlinson-gpu-cluster', 'description': None, 'tags': None, 'properties': {}, 'print_as_yaml': True, 'id': '/subscriptions/fc4f4a6b-4041-4b1c-8249-854d68edcf62/resourceGroups/rapids-deployment-doc/providers/Microsoft.MachineLearningServices/workspaces/skirui-azureml-rapids/computes/jtomlinson-gpu-cluster', 'Resource__source_path': None, 'base_path': '/mnt/batch/tasks/shared/LS_root/mounts/clusters/jtomlinson1/code', 'creation_context': None, 'serialize': <msrest.serialization.Serializer object at 0x7f37edd3dd90>, 'resource_id': None, 'location': 'westus2', 'size': 'STANDARD_NC12S_V3', 'min_instances': 0, 'max_instances': 2, 'idle_time_before_scale_down': 300.0, 'identity': None, 'ssh_public_access_enabled': True, 'ssh_settings': None, 'network_settings': <azure.ai.ml.entities._compute.compute.NetworkSettings object at 0x7f37edd3d850>, 'tier': 'low_priority', 'enable_node_public_ip': True, 'subnet': None})
```

jacobtomlinson replied:
I also don't see RAPIDS mentioned anywhere in this setup. Will this compute cluster have Dask, cudf, etc?

skirui-source replied:
The RAPIDS libraries are installed via the docker container that is used as the environment  to run copies of the training script across the cluster nodes. 

I'm working on adding more details, thanks!

jacobtomlinson commented on source/cloud/azure/azure-ml-multi-node.md:
```diff
@@ -0,0 +1,72 @@
+    name="rapids-docker-image",
+    description="Rapids training Environment")
+
+ml_client.environments.create_or_update(env_docker_image)
```
Once created how do I use this?

jacobtomlinson replied:
Can we include information on what the `Dockerfile` should include?

jacobtomlinson commented on source/cloud/azure/azure-ml-multi-node.md:
```diff
@@ -0,0 +1,72 @@
+
+### Run Example Notebook
+
+Now that we have our environment and custom logic, we can configure and run `command` to submit hyperparameter optimization tuning jobs.
```
I see `azure.ai.ml.command` is mentioned in the notebook in https://github.com/rapidsai/deployment/pull/192/. But this docs page needs to stand on its own too, so we should add an example of running a command here. We should also show how to delete the compute cluster again.

jacobtomlinson commented on source/cloud/azure/azure-ml.md:
```diff
@@ -40,9 +38,12 @@ See the [Azure ML documentation](https://learn.microsoft.com/en-us/azure/machine
 sudo -u azureuser -i <<'EOF'
 
-conda create -y -n rapids {{ rapids_conda_channels }} {{ rapids_conda_packages }} ipykernel
+conda create -y -n rapids -c rapidsai -c conda-forge -c nvidia rapids=23.02 python=3.10 cudatoolkit=11.5 ipykernel
```
Why has this templating been removed?

jacobtomlinson reviewed (commented):
Thanks @skirui-source. Running into a few problems. Here are the steps I followed.

- Open the [build preview](https://rapids-deployment--181.org.readthedocs.build/en/181/cloud/index.html) (which you can find in the checks, click the "readthedocs" check and then "view docs")
- Navigated to "Cloud" -> "Microsoft Azure".
- **Issue 1**: The link to AzureML is broken. I think because the page was renamed.
- Manually navigated to the Azure ML page (because I know what the URL would be) and couldn't see any info on MNMG. Realised this PR adds a second AzureML page.
- **Issue 2**: Having multiple pages could confuse users and this new page isn't linked from anywhere. Please consolidate into a single page. Use section headers so that folks can navigate using the right-hand nav.
- Next I navigated manually to the Azure ML Multi Node page and read the first few paragraphs. Was a little confused about how to achieve the prerequisites. I decided to follow the [single node docs page](https://rapids-deployment--181.org.readthedocs.build/en/181/cloud/azure/azure-ml.html) to get a compute instance.
- Once in Jupyter Lab I created a new notebook with the RAPIDS kernel.
- I followed the instructions in the doc and got to the point of provisioning a compute cluster.
- **Issue 3**: There are missing imports and information.
- **Issue 4**: It's unclear what to do next.
- **Issue 5**: It's unclear how to tear down.
- I then went into the Azure console and deleted the compute cluster and stopped the compute instance.

I think overall this feels incomplete and doesn't have all of the information I needed. There are also various formatting issues that I've commented on.

skirui-source commented:
> Thanks @skirui-source. Running into a few problems. Here are the steps I followed.
> 
> * Open the [build preview](https://rapids-deployment--181.org.readthedocs.build/en/181/cloud/index.html) (which you can find in the checks, click the "readthedocs" check and then "view docs")
> * Navigated to "Cloud" -> "Microsoft Azure".
> * **Issue 1**: The link to AzureML is broken. I think because the page was renamed.
> * ~~Manually navigated to the Azure ML page (because I know what the URL would be) and couldn't see any info on MNMG. Realised this PR adds a second AzureML page.~~
>* ~~**Issue 2**: Having multiple pages could confuse users and this new page isn't linked from anywhere. Please consolidate into a single page. Use section headers so that folks can navigate using the right-hand nav.~~
>* ~~Next I navigated manually to the Azure ML Multi Node page and read the first few paragraphs. Was a little confused about how to achieve the prerequisites. I decided to follow the [single node docs page](https://rapids-deployment--181.org.readthedocs.build/en/181/cloud/azure/azure-ml.html) to get a compute instance.~~
>* ~~Once in Jupyter Lab I created a new notebook with the RAPIDS kernel.~~
> *~~I followed the instructions in the doc and got to the point of provisioning a compute cluster.~~
> * ~~**Issue 3**: There are missing imports and information.~~
> * ~~**Issue 4**: It's unclear what to do next.~~
> * **Issue 5**: It's unclear how to tear down.
> * I then went into the Azure console and deleted the compute cluster and stopped the compute instance.
> 
> I think overall this feels incomplete and doesn't have all of the information I needed. There are also various formatting issues that I've commented on.


I have addressed the above issues, still pending to fix format issues and check build preview (for some reason it still doesn't work) hm 🤔

skirui-source commented on source/cloud/azure/azure-ml-multi-node.md:
```diff
@@ -0,0 +1,72 @@
+
+## Pre-requisites
+
+All you need to get started is an Azure machine learning workspace and development environment must be installed with [Python SDK v2](https://learn.microsoft.com/en-us/python/api/overview/azure/ai-ml-readme?view=azure-python)or [Azure CLI v2](https://learn.microsoft.com/en-us/azure/machine-learning/how-to-configure-cli?tabs=public). You could install Azure Machine Learning on your local computer  ...
```
fixed this issue by merging the separate docs into one 👍🏾

jacobtomlinson commented:
I wonder if the build failure is related to the renaming of the page? Maybe there is a link somewhere that needs updating?

jarmak-nv commented:
Yeah I think the build might be failing because https://github.com/rapidsai/deployment/blob/azureml-multi-node/source/cloud/azure/index.md hasn't been updated to reflect the new name.

jacobtomlinson commented on source/cloud/azure/azureml.md:
````diff
@@ -55,9 +65,147 @@ Launch the instance.
+```python
+from azure.ai.ml.entities import AmlCompute
+
+gpu_compute = AmlCompute(
+    name,
+    type="amlcompute",
+    size,
+    max_instances,
+    idle_time_before_scale_down,
+    tier="low_priority",  # optional
+)
````
This code isn't valid. We should probably stub out the kwargs with some values fo folks to copy/modify.

```suggestion
gpu_compute = AmlCompute(
    name="rapids-example",
    type="amlcompute",
    size="Standard_NC12s_v3",
    max_instances=3,
    idle_time_before_scale_down=300,  # Seconds of idle time before scaling down
    tier="low_priority",  # optional
)
```

jacobtomlinson commented on source/cloud/azure/azureml.md:
```diff
@@ -55,9 +65,147 @@ Launch the instance.
+
+env_docker_image = Environment(
+    build=BuildContext(path="Dockerfile"),
+    name,
+    description,
```
```suggestion
    name="rapids-mlflow",
    description="RAPIDS environment with azureml-mlflow",
```

jacobtomlinson commented on source/cloud/azure/azureml.md:
```diff
@@ -55,9 +65,147 @@ Launch the instance.
+from azure.ai.ml import command, Input
+
+# submit training job
+command_job = command(
+    environment,
+    experiment_name,
+    code,
+    command,
+    inputs,
+    compute,
+)
+returned_job = ml_client.jobs.create_or_update(command_job)
+
+
+# apply hyperparameter sweep_job
+sweep_job = command_job.sweep(
+    compute,
+    sampling_algorithm,
+    primary_metric,
+    goal="Maximize",
+)
```
Again this code is not valid, none of these variables exist and I think they are kwargs rather than args anyway.

We should have a small and complete example here of running a job. Even if it just a small hello world script.

jacobtomlinson commented on source/cloud/azure/azureml.md:
````diff
@@ -55,9 +65,147 @@ Launch the instance.
+### CleanUp
+
+```python
+# Delete amlcompute cluster
+gpu_compute.delete()
+
+# Delete Compute instance
+ml_client.compute.begin_delete("compute_instance_name").wait()
````
I'm seeing the following error when running this line `gpu_compute.delete()`.

```
AttributeError: 'AmlCompute' object has no attribute 'delete'
```

I think it should be

```suggestion
# Delete Compute instance
ml_client.compute.begin_delete(gpu_compute.name).wait()
```

jacobtomlinson reviewed (commented):
Thanks @skirui-source.

I ran through the instructions and it's nearly ready to merge. I think we need a small runnable command example here. See my comments.

skirui-source commented on source/cloud/azure/azureml.md:
```diff
@@ -55,9 +65,165 @@ Launch the instance.
+    environment="rapids-mlflow:1",  # specify version of environment to use
+    experiment_name="test_rapids_mlflow",
+    code=project_folder,
+    command="python train_rapids.py --data_dir ${{inputs.data_dir}} \
```
The `inputs` variable is causing build-docs failures.. i think it assumed that the values should be populated in the `version_template` function as is the case for `{{ rapids_container }}`

Not sure how to resolve this, thoughts?

jacobtomlinson replied:
I've opened #215 to fix this. Once it's merged you'll need to pull in from `main`.

jacobtomlinson reviewed (approved):
Thanks for pulling in more of the example to make this feel complete. It's really difficult to know where the line is in terms of not enough information vs too much information. I think this is about right. Appreciate the back and forth here.
